### PR TITLE
seqGAN

### PR DESCRIPTION
**Pull request by JianGuanTHU**
**Description:**
update seqGAN-tensorflow model.

**Reference Issues:** #XX (XX is the issue number you work on)
#28

### CODE REVIEW

File: contk/dataloader/language_generation.py [hunk 1]
```diff
@@ -111,6 +111,23 @@ def get_inference_metric(self, gen_key="gen"):
 													 gen_key=gen_key))
 		return metric
 
+	def get_gan_metric(self, gen_key="gen_gan"):
```
Comment (kepei1106) [inline contk/dataloader/language_generation.py]: It seems that self-bleu and forward/backward bleu should be added in get_inference_metric. Refer to contk/dataloader/single_turn_dialog.py

File: contk/metric/metric.py [hunk 2]
```diff
@@ -202,6 +205,131 @@ def close(self):
...
+class SelfBleuCorpusMetric(MetricBase):
+	'''Metric for calcualting BLEU.
+
+	Arguments:
+		reference_key (str): Reference sentences are passed to :func:.forward by ``data[reference_key]``.
+			Default: ``resp``.
+		gen_key (str): Sentences generated by model are passed to :func:.forward by
+			``data[gen_prob_key]``. Default: ``gen``.
+	'''
+	def __init__(self, dataloader, gen_key="gen_gan"):
```
Comment (kepei1106) [inline contk/metric/metric.py]: The default value of gen_key doesn't match the hint.

File: contk/metric/metric.py [hunk 3]
```diff
@@ -202,6 +205,131 @@ def close(self):
...
+		self.hyps = []
+
+	def forward(self, data):
+		'''Processing a batch of data.
+
+		Arguments:
+			data (dict): A dict at least contains the following keys.
+			data[reference_key] (list or :class:`numpy.array` of `int`): Reference sentences.
+				Contains start token (eg: ``<go>``) and end token (eg: ``<eos>``).
+				Size: `[batch_size, max_sentence_length]`
```
Comment (kepei1106) [inline contk/metric/metric.py]: self-bleu doesn't use data[reference_key]

**Comment by kepei1106** [comment]
Remove the debug information such as `print` in the next version.

File: contk/metric/metric.py [hunk 4]
```diff
@@ -202,6 +205,131 @@ def close(self):
...
+		reference, fn = ele
+		BLEUscore_f = sentence_bleu(reference, fn, smoothing_function=SmoothingFunction().method1)
+		return BLEUscore_f
+
+	def close(self):
+		'''Return a dict which contains:
+
+			* **bleu**: bleu value.
+		'''
+		SAMPLE = 1000
```
Comment (kepei1106) [inline contk/metric/metric.py]: This hyper-parameter should be the arguments of `__init__` and should be checked to avoid overflow.

File: contk/metric/metric.py (hunk 2 above)
Comment (kepei1106) [reply contk/metric/metric.py]: Other errors of the same kind should be fixed.

**Comment by kepei1106** [comment]
Check whether you use utils/eval_bleu_texygen.py and utils/eval_bleu_concurrent.py. These two files are from other projects. If you really need them, you should pay attention to the open-source license.

File: contk/metric/metric.py [hunk 5]
```diff
@@ -347,6 +348,142 @@ def close(self):
...
+	def close(self):
+		'''Return a dict which contains:
+
+			* **fwbwbleu**: fw/bw bleu value.
+		'''
+		max_len = max([len(self.hyps), len(self.refs)])
+		if self.sample > max_len:
+			self.sample = max_len
+
+		self.sample = 2
```
Comment (kepei1106) [inline contk/metric/metric.py]: What does this operation mean?

File: models/seqGAN-tensorflow/Readme.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+- You should remain similar output in this task.
+
+
+
+### Performance
+
+|        | self-bleu | fwbw-bleu |
+| :----: | :-------: | :-------: |
+| MSCOCO |  0.1569   |  0.2335   |
```
Comment (kepei1106) [inline models/seqGAN-tensorflow/Readme.md]: Are these results self-bleu-4 and bleu-ha-4？Maybe this metric should also support bleu-2/3.

File: models/seqGAN-tensorflow/discriminator.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    def step_decoder(self, generator, session, data, forward_only=False):
+        sentence = []
+        for s in data["sentence"]:
+            sentence.append(np.concatenate((s[1:],[self.data.word2id["<pad>"]] * (self.sequence_length-len(s[1:]))), 0))
+        neg_data = generator.generate(session)
+
+        feed = {
+            self.input_x: np.concatenate((np.array(sentence), neg_data), 0) ,
+            self.input_y: np.concatenate((np.array([[0,1]]*np.shape(sentence)[0]), [[1,0]]*np.shape(neg_data)[0]), 0),
+            self.dropout_keep_prob: self.args.dis_dropout_keep_prob
```
Comment (kepei1106) [inline models/seqGAN-tensorflow/discriminator.py]: Note that dropout_keep_prob=1 when forward_only.

File: contk/metric/metric.py [hunk 8]
```diff
@@ -430,6 +430,153 @@ def close(self):
...
+		'''Processing a batch of data.
+
+		Arguments:
+			data (dict): A dict at least contains the following keys.
+			data[gen_key] (list or :class:`numpy.array` of `int`): Setences generated by model.
+				Contains end token (eg: ``<eos>``), but without start token (eg: ``<go>``).
+				Size: `[batch_size, gen_sentence_length]`.
+		'''
+		gen = data[self.gen_key]
+		resp = self.dataloader.data["test"]["sen"]
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: use XXX_key instead of "sen"

File: contk/metric/metric.py [hunk 9]
```diff
@@ -430,6 +430,153 @@ def close(self):
...
+			return result
+		except ZeroDivisionError as _:
+			raise ZeroDivisionError("Bleu smoothing divided by zero. This is a known bug of corpus_bleu, \
+				usually caused when there is only one sample and the sample length is 1.")
+
+class FwBwBleuCorpusMetric(MetricBase):
+	'''Metric for calcualting BLEU.
+
+	Arguments:
+		reference_key (str): Reference sentences are passed to :func:.forward by ``data[reference_key]``.
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: use reference_allvocabs    check other metric changes

File: contk/metric/metric.py [hunk 10]
```diff
@@ -430,6 +430,153 @@ def close(self):
...
+			Default: ``resp``.
+		gen_key (str): Sentences generated by model are passed to :func:.forward by
+			``data[gen_prob_key]``. Default: ``gen``.
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: gen_key not gen_prob_key

File: contk/metric/metric.py [hunk 11]
```diff
@@ -430,6 +430,160 @@ def close(self):
...
+				usually caused when there is only one sample and the sample length is 1.")
+
+class FwBwBleuCorpusMetric(MetricBase):
+	'''Metric for calcualting BLEU.
+
+	Arguments:
+		dataloader (:class:contk.BasicLanguageGeneration): A language generation dataloader.
+		reference_test_key (str): Reference sentences with all vocabs in test data
+			are passed to :func:.forward by ``data[reference_test_key]``.
+			Default: ``sen``.
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: fix this

File: contk/metric/metric.py [hunk 12]
```diff
@@ -430,6 +430,160 @@ def close(self):
...
+		gen_key (str): Sentences generated by model are passed to :func:.forward by
+			``data[gen_key]``. Default: ``gen``.
+		sample (int): Number of samples sampled from the generated sentences. Default: 1000.
+	'''
+	def __init__(self, dataloader,
+			reference_test_key="sent",
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: use sent_allvocabs. please refer to other metrics

File: tests/models/dummy_glove_300d.txt [hunk 13]
```diff
@@ -0,0 +1,3 @@
....082966 0.16626 -0.10601 0.35861 0.16922 0.07259 -0.24803 -0.10024 -0.52491 -0.17745 -0.36647 0.2618 -0.012077 0.08319 -0.21528 0.41045 0.29136 0.30869 0.078864 0.32207 -0.041023 -0.1097 -0.092041 -0.12339 -0.16416 0.35382 -0.082774 0.33171 -0.24738 -0.048928 0.15746 0.18988 -0.026642 0.063315 -0.010673 0.34089 1.4106 0.13417 0.28191 -0.2594 0.055267 -0.052425 -0.25789 0.019127 -0.022084 0.32113 0.068818 0.51207 0.16478 -0.20194 0.29232 0.098575 0.013145 -0.10652 0.1351 -0.045332 0.20697 -0.48425 -0.44706 0.0033305 0.0029264 -0.10975 -0.23325 0.22442 -0.10503 0.12339 0.10978 0.048994 -0.25157 0.40319 0.35318 0.18651 -0.023622 -0.12734 0.11475 0.27359 -0.21866 0.015794 0.81754 -0.023792 -0.85469 -0.16203 0.18076 0.028014 -0.1434 0.0013139 -0.091735 -0.089704 0.11105 -0.16703 0.068377 -0.087388 -0.039789 0.014184 0.21187 0.28579 -0.28797 -0.058996 -0.032436 -0.0047009 -0.17052 -0.034741 -0.11489 0.075093 0.099526 0.048183 -0.073775 -0.41817 0.0041268 0.44414 -0.16062 0.14294 -2.2628 -0.027347 0.81311 0.77417 -0.25639 -0.11576 -0.11982 -0.21363 0.028429 0.27261 0.031026 0.096782 0.0067769 0.14082 -0.013064 -0.29686 -0.079913 0.195 0.031549 0.28506 -0.087461 0.0090611 -0.20989 0.053913
```
Comment (hzhwcmhf) [inline tests/models/dummy_glove_300d.txt]: Don't change this.
